cmd/contour/servecontext.go

```diff
@@ -0,0 +1,200 @@
+// Copyright © 2019 Heptio
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"crypto/rand"
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/heptio/contour/internal/contour"
+)
+
+type serveContext struct {
+	// contour's kubernetes client parameters
+	InCluster  bool   `yaml:"incluster,omitempty"`
+	Kubeconfig string `yaml:"kubeconfig,omitempty"`
+
+	// contour's xds service parameters
+	xdsAddr                         string
+	xdsPort                         int
+	caFile, contourCert, contourKey string
+
+	// contour's debug handler parameters
+	debugAddr string
+	debugPort int
+
+	// contour's metrics handler parameters
+	metricsAddr string
+	metricsPort int
+
+	// ingressroute root namespaces
+	rootNamespaces string
+
+	// ingress class
+	ingressClass string
+
+	// envoy's stats listener parameters
+	statsAddr string
+	statsPort int
+
+	// envoy's listener parameters
+	useProxyProto bool
+
+	// envoy's http listener parameters
+	httpAddr      string
+	httpPort      int
+	httpAccessLog string
+
+	// envoy's https listener parameters
+	httpsAddr      string
+	httpsPort      int
+	httpsAccessLog string
+
+	// PermitInsecureGRPC disables TLS on Contour's gRPC listener.
+	PermitInsecureGRPC bool `yaml:"-"`
+
+	TLSConfig `yaml:"tls,omitempty"`
+
+	// DisablePermitInsecure disables the use of the
+	// permitInsecure field in IngressRoute.
+	DisablePermitInsecure bool `yaml:"disablePermitInsecure,omitempty"`
+
+	EnableLeaderElection bool
+	LeaderElectionConfig `yaml:"leaderelection,omitempty"`
+}
+
+// newServeContext returns a serveContext initialized to defaults.
+// Supply one or more functions to modify the defaults.
+func newServeContext(options ...func(*serveContext)) *serveContext {
```

Comment:
I see what you're doing here passing in a function to change the fields in serveContext for testing, but its odd that this will leak into non text contexts.

I suggest something simpler. The test is only interested in does yaml decoding affect fields which are public -- you don't need to set the ones that yaml can't touch. Rather than passing in a option func style try something like this

want := func() *serveContext {
ctx := newServeContext()
ctx.foo = 1
ctx.bar =2
return ctx
}()

Reply:
Okay, I see. That is simpler, thanks.